### ISSUE TITLE
problems with empty non-string values in csv conversion

### ISSUE THREAD

**Issue by massimomorara** [danielaparker/jsoncons#62]
Sorry again for my bad English.
I don't know if I doing something wrong, but... I have a problem with values imported for non-string types from empty csv fields. 
I've written a minimal program to show the problem.

[test-001.cpp.txt](https://github.com/danielaparker/jsoncons/files/140858/test-001.cpp.txt)

In a nutshell...
Using the following csv content

`"bool-f,int-f,float-f,string-f"`
`"\n,,,,"`
`"\ntrue,12,24.7,\"test string\","`
`"\n,,,,"`

(a row of empty values, a row with valid values, another row with empty values) with the following data types string

`"boolean,integer,float,string"`

I obtain that:
- the empty boolean field are ever ignored (like when "ignore_empty_values" is set)
- the imported integer value for the first row is a randomic big number (different on every run of the program); for the third row is the same value (12) of the second row;
- the imported float value for both first and third row is a randomic little value (different for the first and the third; different on every run of the program)

An example of output

`[`
`    {`
`        "float-f":6.952914947851373e-310,`
`        "int-f":140728565250248,`
`        "string-f":""`
`    },`
`    {`
`        "bool-f":true,`
`        "float-f":24.7,`
`        "int-f":12,`
`        "string-f":"test string"`
`    },`
`    {`
`        "float-f":9.99944203648299e-317,`
`        "int-f":12,`
`        "string-f":""`
`    }`
`]`

My environment is a Debian Jessie amd64 platform; I'm using clang++ 3.5.0 with the following compilation command

`clang++ -o test-001.o -c -g -Wall -ansi -pedantic -std=c++11 -stdlib=libc++ -I. test-001.cpp`

and the following linking command

`clang++ -o test-001 -std=c++11 -stdlib=libc++ test-001.o`

**Comment by massimomorara** [comment]
I've tried with g++ too (version 4.9.2) and the problem is a little different:
- the imported empty integer values are ever 5 (first and third row; every run of the program)
- the imported empty float values are equal in the first and in the third row bat different in every run.

An example of output

`[`
`    {`
`        "float-f":6.952663267993231e-310,`
`        "int-f":5,`
`        "string-f":""`
`    },`
`    {`
`        "bool-f":true,`
`        "float-f":24.7,`
`        "int-f":12,`
`        "string-f":"test string"`
`    },`
`    {`
`        "float-f":6.952663267993231e-310,`
`        "int-f":5,`
`        "string-f":""`
`    }`
`]`

**Comment by massimomorara** [comment]
I'm not sure about the intended behaviour, but I suppore that the problem is in "basic_csv_parser::end_value()" (`csv_parser.hpp`) where `val` isn't initialized in cases "`data_types::integer_t`" and "`data_types::float_t`".
I suppose that can be enough inizialize

`long long val = 0LL;`

and 

`double val = 0.0;`

But i think is better to check the length of `string_buffer_`; something like

`long long val = 0LL;`
`if ( false == string_buffer_.empty() )`
`std::istringstream(string_buffer_) >> val;`
`handler_->value(val, *this);`

For boolean, if "`false`" is a good default (empty) value, i suppose that the `data_types::boolean_t` can be rewritten (should work with wchar_t too)

`            case data_types::boolean_t:`
`                  handler_->value(`
`                     (   (   (1 == string_buffer_.length())`
`                          && ('1' == string_buffer_[0]))`
`                      || (   (4 == string_buffer_.length())`
`                          && (CharT('t') == std::tolower(string_buffer_[0], std::locale()))`
`                          && (CharT('r') == std::tolower(string_buffer_[1], std::locale()))`
`                          && (CharT('u') == std::tolower(string_buffer_[2], std::locale()))`
`                          && (CharT('e') == std::tolower(string_buffer_[3], std::locale())))), *this);`
`                break;`

**Comment by danielaparker** [comment]
I've aded a property default_values to csv_parameters, which allows specifying default values for columns when values are bad, or in the case of strings, empty. The default defaults are null for integer, float and bool, and empty string for string.

With no defaults specified, your example now produces:

    [
        {
            "bool-f":null,
            "float-f":null,
            "int-f":null,
            "string-f":""
        },
        {
            "bool-f":true,
            "float-f":24.7,
            "int-f":12,
            "string-f":"test string"
        },
        {
            "bool-f":null,
            "float-f":null,
            "int-f":null,
            "string-f":""
        }
    ]

With defaults specified, e.g.

    params.default_values("false,0,0.0,");

it now produces:
    [
        {
            "bool-f":false,
            "float-f":0.0,
            "int-f":0,
            "string-f":""
        },
        {
            "bool-f":true,
            "float-f":24.7,
            "int-f":12,
            "string-f":"test string"
        },
        {
            "bool-f":false,
            "float-f":0.0,
            "int-f":0,
            "string-f":""
        }
    ]

Daniel

**Comment by massimomorara** [comment]
Nice solution.
Only one doubt: for uniformity, it would not be better if, by default, even the string values were null?

**Comment by danielaparker** [comment]
It's a valid point, but the difference is that a zero length string is still a valid string, unlike the other cases. And in practice I think we (or at least I) would almost always want a zero length string to be a string rather than a null.

Also to note: I've deprecated the csv_parameter methods that take delimited strings, including header, data_types, and default_values. These have been replaced with column_names, column_types, and column_defaults that take initializer lists of strings. So instead of

    params.default_values("false,0,0.0,");

use

    params.default_values({"false","0","0.0",""});

(The old methods can still be used as long as JSONCONS_NO_DEPRECATED is not set.)
